path/dynamic/dstarlite.go

```diff
@@ -0,0 +1,462 @@
+	   {09”} else if (g(u) = rhs(u) AND u ∈ U) U.Remove(u);
+	*/
+	uid := u.ID()
+	inQueue := d.queue.has(uid)
+	switch {
+	case inQueue && u.g != u.rhs:
+		d.queue.update(uid, d.keyFor(u))
+	case !inQueue && u.g != u.rhs:
+		d.queue.insert(u, d.keyFor(u))
+	case inQueue && u.g == u.rhs:
+		d.queue.remove(uid)
+	}
+}
+
+// findShortestPath is the ComputeShortestPath procedure in the D* Lite papers.
+func (d *DStarLite) findShortestPath() {
+	/*
+	   procedure ComputeShortestPath()
+	   {10”} while (U.TopKey() < CalculateKey(s_start) OR rhs(s_start) > g(s_start))
+	   {11”} u = U.Top();
+	   {12”} k_old = U.TopKey();
+	   {13”} k_new = CalculateKey(u);
+	   {14”} if(k_old < k_new)
+	   {15”}   U.Update(u, k_new);
+	   {16”} else if (g(u) > rhs(u))
+	   {17”}   g(u) = rhs(u);
+	   {18”}   U.Remove(u);
+	   {19”}   for all s ∈ Pred(u)
+	   {20”}     if (s != s_goal) rhs(s) = min(rhs(s), c(s, u) + g(u));
+	   {21”}     UpdateVertex(s);
+	   {22”} else
+	   {23”}   g_old = g(u);
+	   {24”}   g(u) = ∞;
+	   {25”}   for all s ∈ Pred(u) ∪ {u}
+	   {26”}     if (rhs(s) = c(s, u) + g_old)
+	   {27”}       if (s != s_goal) rhs(s) = min s'∈Succ(s)(c(s, s') + g(s'));
+	   {28”}     UpdateVertex(s);
+	*/
+	for d.queue.Len() != 0 { // We use d.queue.Len since d.queue does not return an infinite key when empty.
+		u := d.queue.top()
+		if !u.key.less(d.keyFor(d.s)) && d.s.rhs <= d.s.g {
+			break
+		}
+		switch kNew := d.keyFor(u); {
+		case u.key.less(kNew):
+			d.queue.update(u.ID(), kNew)
+		case u.g > u.rhs:
+			u.g = u.rhs
+			d.queue.remove(u.ID())
+			for _, _s := range d.world.To(u) {
+				s := _s.(*dStarLiteNode)
+				if s.ID() != d.t.ID() {
+					s.rhs = math.Min(s.rhs, d.weight(d.world.Edge(s, u))+u.g)
+				}
+				d.update(s)
+			}
+		default:
+			gOld := u.g
+			u.g = math.Inf(1)
+			for _, _s := range append(d.world.To(u), u) {
+				s := _s.(*dStarLiteNode)
+				// This is necessary since we are potentially
+				// asking for the weight of u to u, but we
+				// never return a self-edge.
+				var w float64
+				if s.ID() != u.ID() {
+					w = d.weight(d.world.Edge(s, u))
+				}
+				if s.rhs == w+gOld {
+					if s.ID() != d.t.ID() {
+						s.rhs = math.Inf(1)
+						for _, t := range d.world.From(s) {
+							s.rhs = math.Min(s.rhs, d.weight(d.world.Edge(s, t))+t.(dStarLiteNode).g)
+						}
+					}
+				}
+				d.update(s)
+			}
+		}
+	}
+}
+
+// Step performs one movement step along the best path towards the goal.
+// It returns false if no further progression toward the goal can be
+// achieved, either because the goal has been reached or because there
+// is no path.
+func (d *DStarLite) Step() bool {
+	/*
+	   procedure Main()
+	   {32”} while (s_start != s_goal)
+	   {33”} // if (g(rhs_start) = ∞) then there is no known path
+	   {34”}   s_start = argmin s'∈Succ(s_start)(c(s_start, s') + g(s'));
+	*/
+	if d.s.ID() == d.t.ID() {
+		return false
+	}
+	if math.IsInf(d.s.rhs, 1) {
+		return false
+	}
+	min := math.Inf(1)
```

Comment:
I wanted to make the distinction that the move is not being done by the algorithm - this was the source of confusion before. AFAICS the English comment indicates that it is the movement of the robot, not the change of the algorithm's representation of location; so the action described by `{35”}` takes place after the return true with (in the tests) `l.MoveTo(d.Here())`.